File: src/partial_eval.c
#include "ast.h"
#include "defs.h"

void partial_eval(ASTNode *node) {
  switch (node->token) {
  case Neg:
    partial_eval(node->lhs);
    if (node->lhs->token == Fixnum) {
      node->token = Fixnum;
      node->value = -node->lhs->value;
      node->lhs = 0;
    }
    break;
  case Add:
    partial_eval(node->lhs);
    partial_eval(node->rhs);
    if (node->lhs->token == Fixnum && node->rhs->token == Fixnum) {
      node->token = Fixnum;
      node->value = node->lhs->value + node->rhs->value;
      node->lhs = 0;
      node->rhs = 0;
    }
    break;
  default:
    break;
  }
}

File: src/pass/partial_eval.rs
use crate::ast::Node;

pub fn partial_eval(node: Box<Node>) -> Box<Node> {
    let node = match *node {
        Node::Program(sub_node) => Node::Program(partial_eval(sub_node)),
        Node::Neg(sub_node) => {
            let sub_node = partial_eval(sub_node);
            if let Node::Fixnum(num) = *sub_node {
                Node::Fixnum(-num)
            } else {
                Node::Neg(sub_node)
            }
        }
        Node::Add(lhs, rhs) => {
            let lhs = partial_eval(lhs);
            let rhs = partial_eval(rhs);
            if lhs.fixnum().is_some() && rhs.fixnum().is_some() {
                Node::Fixnum(lhs.fixnum().unwrap() + rhs.fixnum().unwrap())
            } else {
                Node::Add(lhs, rhs)
            }
        }
        _ => *node,
    };
    Box::new(node)
}
